components/brave_rewards/renderer/rewards_render_frame_observer.cc

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_rewards/renderer/rewards_render_frame_observer.h"
+
+#include <array>
+#include <optional>
+#include <utility>
+
+#include "base/feature_list.h"
+#include "base/functional/bind.h"
+#include "brave/components/brave_rewards/common/features.h"
+#include "brave/components/brave_rewards/common/publisher_utils.h"
+#include "brave/components/brave_rewards/resources/grit/creator_detection_generated.h"
+#include "content/public/renderer/render_frame.h"
+#include "gin/function_template.h"
+#include "third_party/blink/public/common/associated_interfaces/associated_interface_provider.h"
+#include "third_party/blink/public/platform/platform.h"
+#include "third_party/blink/public/platform/scheduler/web_agent_group_scheduler.h"
+#include "third_party/blink/public/web/blink.h"
+#include "third_party/blink/public/web/web_local_frame.h"
+#include "third_party/blink/public/web/web_script_source.h"
+
+namespace brave_rewards {
+
+namespace {
+
+v8::Isolate* GetIsolate(content::RenderFrame* render_frame) {
+  DCHECK(render_frame);
+  return render_frame->GetWebFrame()->GetAgentGroupScheduler()->Isolate();
+}
+
+v8::Local<v8::Context> GetMainWorldContext(content::RenderFrame* render_frame) {
+  DCHECK(render_frame);
+  return render_frame->GetWebFrame()->MainWorldScriptContext();
+}
+
+std::optional<std::string> GetDetectionScript(
+    content::RenderFrame* render_frame) {
+  if (!render_frame || !render_frame->IsMainFrame()) {
+    return std::nullopt;
+  }
+
+  auto* web_frame = render_frame->GetWebFrame();
+  DCHECK(web_frame);
+  if (web_frame->IsProvisional()) {
+    return std::nullopt;
+  }
+
+  GURL origin = url::Origin(web_frame->GetSecurityOrigin()).GetURL();
+  if (!origin.is_valid() || !origin.SchemeIs(url::kHttpsScheme) ||
+      !IsMediaPlatformURL(origin)) {
+    return std::nullopt;
+  }
+
+  // TODO(zeparsing): Do we need to run this on a separate thread? Do we need to
+  // optimize the loading of this string (or its JS compilation)?
+  return blink::Platform::Current()->GetDataResourceString(
```

Comment:
If you load the resource as a raw string it's a memory mapped from .pak file => about nothing to optimize here.

Optimizing JS compilation is very hard for such custom injects. The compilation results for one renderer usually can't be reused for other for sec reasons. Therefore, we don't use such optimization for other features and it's definitely another complex task.